mayastor/src/bdev/nexus/nexus_child_error_store.rs

```diff
@@ -0,0 +1,302 @@
+use spdk_sys::{spdk_bdev, spdk_bdev_io_type};
+use std::{
+    fmt::{Debug, Display},
+    time::{Duration, Instant},
+};
+
+use crate::subsys::Config;
+
+use crate::bdev::nexus::nexus_io::{io_status, io_type};
+
+use crate::core::{Cores, Reactors};
+
+use crate::bdev::nexus::{
+    nexus_bdev,
+    nexus_bdev::{
+        nexus_lookup,
+        Error::{ChildMissing, ChildMissingErrStore},
+        Nexus,
+    },
+};
+
+use serde::export::{fmt::Error, Formatter};
+
+#[derive(Copy, Clone)]
+pub struct NexusChildErrorRecord {
+    io_offset: u64,
+    io_num_blocks: u64,
+    timestamp: Instant,
+    io_error: i32,
+    io_op: spdk_bdev_io_type,
+}
+
+impl Default for NexusChildErrorRecord {
+    fn default() -> Self {
+        Self {
+            io_op: 0,
+            io_error: 0,
+            io_offset: 0,
+            io_num_blocks: 0,
+            timestamp: Instant::now(), // for want of another suitable default
+        }
+    }
+}
+
+pub struct NexusErrStore {
+    no_of_records: usize,
+    next_record_index: usize,
+    records: Vec<NexusChildErrorRecord>,
+}
+
+impl NexusErrStore {
+    pub const READ_FLAG: u32 = 1;
+    pub const WRITE_FLAG: u32 = 2;
+    pub const UNMAP_FLAG: u32 = 4;
+    pub const FLUSH_FLAG: u32 = 8;
+    pub const RESET_FLAG: u32 = 16;
+
+    pub const IO_FAILED_FLAG: u32 = 1;
+
+    // the following definitions are for the error_store unit test
+    pub const IO_TYPE_READ: u32 = io_type::READ;
+    pub const IO_TYPE_WRITE: u32 = io_type::WRITE;
+    pub const IO_TYPE_UNMAP: u32 = io_type::UNMAP;
+    pub const IO_TYPE_FLUSH: u32 = io_type::FLUSH;
+    pub const IO_TYPE_RESET: u32 = io_type::RESET;
+
+    pub const IO_FAILED: i32 = io_status::FAILED;
+
+    pub fn new(max_records: usize) -> Self {
+        let mut es = NexusErrStore {
+            no_of_records: 0,
+            next_record_index: 0,
+            records: Vec::with_capacity(max_records),
+        };
+
+        let er: NexusChildErrorRecord = Default::default();
+
+        for _ in 0 .. max_records {
+            es.records.push(er);
+        }
+        es
+    }
+
+    pub fn add_record(
+        &mut self,
+        io_op: spdk_bdev_io_type,
+        io_error: i32,
+        io_offset: u64,
+        io_num_blocks: u64,
+        timestamp: Instant,
+    ) {
+        let new_record = NexusChildErrorRecord {
+            io_op,
+            io_error,
+            io_offset,
+            io_num_blocks,
+            timestamp,
+        };
+
+        self.records[self.next_record_index] = new_record;
+
+        if self.no_of_records < self.records.len() {
+            self.no_of_records += 1;
+        };
+        self.next_record_index =
+            (self.next_record_index + 1) % self.records.len();
+        // debug!("added record - buffer is {}", &self);
```

Comment:
did you mean to keep this commented?

Reply:
I did. It makes a nice list of the records when uncommented, but we don't necessarily want that in the output. I also didn't want to forget how to do it, if needed.